src/plugin/ipc/socket/socketconnection.cpp

```diff
@@ -381,6 +386,38 @@ void TcpConnection::drain()
     markExternalConnect();
   }
 
+  // Non blocking connect; need to hang around until it is writable.
+  if (_type == TCP_CONNECT_IN_PROGRESS) {
+    fd_set wfds;
+    struct timeval tv;
+    int retval;
+
+    FD_ZERO(&wfds);
+    FD_SET(_fds[0], &wfds);
+
+    tv.tv_sec = 60;
+    tv.tv_usec = 0;
+
+    // FIXME: Replace select with poll/epoll (to handle fd > 1024).
+    retval = select(_fds[0] + 1, NULL, &wfds, NULL, &tv);
+    /* Don't rely on the value of tv now! */
+
+    if (retval == -1) {
+      JTRACE("select() failed") (JASSERT_ERRNO);
+    } else if (FD_ISSET(_fds[0], &wfds)) {
+      int val = -1;
+      socklen_t sz = sizeof(val);
+      getsockopt(_fds[0], SOL_SOCKET, SO_ERROR, &val, &sz);
+      JTRACE("Connect-in-progress socket is now writable.") (_fds[0]);
+      _type = TCP_CONNECT;
+    } else {
+      JWARNING(false) (_fds[0])
+        .Text("Failed to wait for EINPROGRESS connect to return.\n"
+              "Marking the socket as external and continuing.");
```

Comment:
@karya0:  I propose this more verbose explanation for the `JWARNING`.  It will help users unfamiliar with the internals of DMTCP.
```
       .Text("connect() returned EINPROGRESS.  Socket still not writable\n"
             "after 60 seconds.   The socket is probably connected to an\n"
             "external process not under DMTCP control.\n"
             "Marking this socket as external and continuing to checkpoint.");
```